webpack/worker/worker_adapter.js

```diff
@@ -10,15 +10,68 @@
  * https://medium.com/@mmorearty/how-to-create-a-persistent-worker-for-bazel-7738bba2cabb
  * for more background on the worker protocol.
  */
+
 const worker = require('@bazel/worker')
-const cp = require('child_process')
 const fs = require('fs')
+const path = require('path')
+const WebpackCLI = require('webpack-cli')
+const Linker = require(path.join(process.cwd(), process.env._LINKER_PATH))
+const Runfiles = require(process.env.BAZEL_NODE_RUNFILES_HELPER)
 
-const webpackCli = require('webpack-cli')
+class WorkerAwareCLI extends WebpackCLI {
+  /** @type {import("webpack").Compiler | null} */
+  compiler = null
 
-const MNEMONIC = 'webpack'
+  /** @type {{[k: string]: unknown} | null} */
+  options = null
+
+  /** @type {Function | null} */
+  resolve = null
+  /** @type {Function | null} */
+  reject = null
+
+  /**
+   *
+   * @param {import("webpack").StatsError} err
+   * @param {import("webpack").Stats} stats
+   */
+  callback(err, stats) {
+    if (err && this.reject) {
+      console.err(err)
+      this.reject(err)
+    } else if (!err && this.resolve) {
+      worker.log(stats.toString()) 
+      this.resolve(true)
+    }
+  }
+
+  async discard() {
+    await new Promise((resolve, reject) => this.compiler.close(e => {
+      if (e) {
+        return reject(e)
+      }
+      resolve()
+    }))
+    this.compiler = null
+    this.options = null
+  }
 
-/** @typedef {{type: "built" | "error"}} IPCMessage */
+  async createCompiler(options) {
+    if (this.options != null && JSON.stringify(options) != JSON.stringify(this.options)) {
+      worker.log(`[${MNEMONIC}] options have changed. discarding webpack cache.`)
+      await this.discard(); 
+    }
+    const cb = (err, stats) => this.callback(err, stats)
```

Comment:
Why not pass in this.callback directly? I'm not sure that creating the `cb` anonymous function does anything

Reply:
Here it preserves `this` scope. if I just pass `this.callback`, the scope of `this` would be an internal webpack class.  an alternative to this is `this.callback.bind(this)` which I don't like.